src/Aspire.Hosting.Testing/DistributedApplicationExtensions.cs

```diff
@@ -0,0 +1,141 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using Aspire.Hosting.ApplicationModel;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Aspire.Hosting.Testing;
+
+/// <summary>
+/// Extensions for working with <see cref="DistributedApplication"/> in test code.
+/// </summary>
+public static class DistributedApplicationExtensions
+{
+    /// <summary>
+    /// Creates an <see cref="HttpClient"/> configured to communicate with the specified resource.
+    /// </summary>
+    /// <param resourceName="app">The application.</param>
+    /// <param resourceName="resourceName">The resourceName of the resource.</param>
+    /// <param resourceName="endpointName">The resourceName of the endpoint on the resource to communicate with.</param>
+    /// <returns>The <see cref="HttpClient"/>.</returns>
+    public static HttpClient CreateHttpClient(this DistributedApplication app, string resourceName, string? endpointName = default)
+    {
+        var applicationModel = app.Services.GetRequiredService<DistributedApplicationModel>();
+
+        var resources = applicationModel.Resources;
+        var resource = resources.FirstOrDefault(r => string.Equals(r.Name, resourceName, StringComparison.OrdinalIgnoreCase));
+
+        if (resource is null)
+        {
+            throw new ArgumentException($"Resource '{resourceName}' not found", nameof(resourceName));
+        }
+
+        if (!resource.TryGetAllocatedEndPoints(out var endpoints))
+        {
+            throw new InvalidOperationException($"Cannot create a client for resource '{resourceName}' because it has no allocated endpoints.");
+        }
+
+        AllocatedEndpointAnnotation? endpoint = null;
+
+        if (!string.IsNullOrEmpty(endpointName))
+        {
+            endpoint = endpoints.FirstOrDefault(e => string.Equals(e.Name, endpointName, StringComparison.OrdinalIgnoreCase));
```

Comment:
SingleOrDefault. If we have an endpoint with the same name then we should consider it an exception.

Reply:
We should catch that during app model build / launch, ideally. I'll add a safeguard here anyway